File: go/activation.go
package zymeup

type ActivationClient struct {
	client *Client
}

func NewActivationClient(c *Client) *ActivationClient {
	return &ActivationClient{client: c}
}

func (c *ActivationClient) ListProviders(params map[string]interface{}) ([]map[string]interface{}, error) {
	path := "/api/v1/marketplace/providers" + buildQueryString(params)
	resp, err := c.client.doRequest("GET", path, nil)
	if err != nil {
		return nil, err
	}
	var result struct {
		Code int `json:"code"`
		Data struct {
			Items []map[string]interface{} `json:"items"`
			Total int                      `json:"total"`
		} `json:"data"`
	}
	if err := decodeResponse(resp, &result); err != nil {
		return nil, err
	}
	return result.Data.Items, nil
}

func (c *ActivationClient) GetProvider(slug string) (map[string]interface{}, error) {
	resp, err := c.client.doRequest("GET", "/api/v1/marketplace/providers/"+slug, nil)
	if err != nil {
		return nil, err
	}
	var result struct {
		Code int                    `json:"code"`
		Data map[string]interface{} `json:"data"`
	}
	if err := decodeResponse(resp, &result); err != nil {
		return nil, err
	}
	return result.Data, nil
}

func (c *ActivationClient) List() ([]map[string]interface{}, error) {
	resp, err := c.client.doRequest("GET", "/api/v1/marketplace/activations", nil)
	if err != nil {
		return nil, err
	}
	var result struct {
		Code int `json:"code"`
		Data struct {
			Items []map[string]interface{} `json:"items"`
			Total int                      `json:"total"`
		} `json:"data"`
	}
	if err := decodeResponse(resp, &result); err != nil {
		return nil, err
	}
	return result.Data.Items, nil
}

func (c *ActivationClient) Get(id string) (map[string]interface{}, error) {
	resp, err := c.client.doRequest("GET", "/api/v1/marketplace/activations/"+id, nil)
	if err != nil {
		return nil, err
	}
	var result struct {
		Code int                    `json:"code"`
		Data map[string]interface{} `json:"data"`
	}
	if err := decodeResponse(resp, &result); err != nil {
		return nil, err
	}
	return result.Data, nil
}

func (c *ActivationClient) Activate(data map[string]interface{}) (map[string]interface{}, error) {
	resp, err := c.client.doRequest("POST", "/api/v1/marketplace/activations", data)
	if err != nil {
		return nil, err
	}
	var result struct {
		Code int                    `json:"code"`
		Data map[string]interface{} `json:"data"`
	}
	if err := decodeResponse(resp, &result); err != nil {
		return nil, err
	}
	return result.Data, nil
}

func (c *ActivationClient) Pause(id string) error {
	resp, err := c.client.doRequest("POST", "/api/v1/marketplace/activations/"+id+"/pause", nil)
	if err != nil {
		return err
	}
	return decodeResponse(resp, nil)
}

func (c *ActivationClient) Resume(id string) error {
	resp, err := c.client.doRequest("POST", "/api/v1/marketplace/activations/"+id+"/resume", nil)
	if err != nil {
		return err
	}
	return decodeResponse(resp, nil)
}

func (c *ActivationClient) Revoke(id string) error {
	resp, err := c.client.doRequest("POST", "/api/v1/marketplace/activations/"+id+"/revoke", nil)
	if err != nil {
		return err
	}
	return decodeResponse(resp, nil)
}

File: rust/src/activation.rs
use reqwest::Method;

use crate::error::Result;
use crate::http_client::HttpClient;
use crate::types::ProviderActivation;

pub struct ActivationClient {
    pub(crate) inner: HttpClient,
}

impl ActivationClient {
    pub fn new(inner: HttpClient) -> Self {
        Self { inner }
    }
}

impl ActivationClient {
    pub async fn list_providers(
        &self,
        capability: Option<&str>,
    ) -> Result<serde_json::Value> {
        let mut path = "/api/v1/marketplace/providers".to_string();
        if let Some(c) = capability {
            path.push_str(&format!("?capability={}", urlencoding::encode(c)));
        }
        self.inner.request(&path, Method::GET, None).await
    }

    pub async fn get_provider(&self, slug: &str) -> Result<serde_json::Value> {
        self.inner
            .request(
                &format!("/api/v1/marketplace/providers/{}", slug),
                Method::GET,
                None,
            )
            .await
    }

    pub async fn list_activations(
        &self,
    ) -> Result<serde_json::Value> {
        self.inner
            .request("/api/v1/marketplace/activations", Method::GET, None)
            .await
    }

    pub async fn get_activation(&self, id: &str) -> Result<ProviderActivation> {
        self.inner
            .request(
                &format!("/api/v1/marketplace/activations/{}", id),
                Method::GET,
                None,
            )
            .await
    }

    pub async fn activate(
        &self,
        body: serde_json::Value,
    ) -> Result<ProviderActivation> {
        self.inner
            .request(
                "/api/v1/marketplace/activations",
                Method::POST,
                Some(body),
            )
            .await
    }

    pub async fn pause(&self, id: &str) -> Result<serde_json::Value> {
        self.inner
            .request(
                &format!("/api/v1/marketplace/activations/{}/pause", id),
                Method::POST,
                Some(serde_json::json!({})),
            )
            .await
    }

    pub async fn resume(&self, id: &str) -> Result<serde_json::Value> {
        self.inner
            .request(
                &format!("/api/v1/marketplace/activations/{}/resume", id),
                Method::POST,
                Some(serde_json::json!({})),
            )
            .await
    }

    pub async fn revoke(
        &self,
        id: &str,
        reason: Option<&str>,
    ) -> Result<serde_json::Value> {
        let body = match reason {
            Some(r) => serde_json::json!({ "reason": r }),
            None => serde_json::json!({}),
        };
        self.inner
            .request(
                &format!("/api/v1/marketplace/activations/{}/revoke", id),
                Method::POST,
                Some(body),
            )
            .await
    }
}
